File: java-sdks/src/main/java/com/suisdks/sui/grpc/SuiGrpcClient.java
package com.suisdks.sui.grpc;

import java.util.List;
import java.util.Map;
import java.util.Arrays;
import java.util.Objects;
import java.util.stream.Collectors;

public final class SuiGrpcClient implements AutoCloseable {
    private final GrpcTransport transport;

    public SuiGrpcClient(GrpcTransport transport) {
        this.transport = transport;
    }

    public static SuiGrpcClient fromOfficialGrpc(String target, boolean plaintext) {
        return new SuiGrpcClient(new OfficialGrpcTransport(target, plaintext));
    }

    public Object call(String method, List<Object> params) {
        GrpcResponse r = transport.unary(new GrpcRequest(method, params));
        if (r.hasError()) {
            throw new IllegalStateException("gRPC error: " + r.error());
        }
        return r.result();
    }

    public Object getObject(String objectId, Map<String, Object> options) {
        return call("sui_getObject", List.of(objectId, options));
    }

    public Object executeTransactionBlock(String txBytesBase64, List<String> signatures, Map<String, Object> options) {
        return call("sui_executeTransactionBlock", List.of(txBytesBase64, signatures, options));
    }

    public Object dryRunTransactionBlock(String txBytesBase64) {
        return call("sui_dryRunTransactionBlock", List.of(txBytesBase64));
    }

    public Object devInspectTransactionBlock(String sender, Object txData) {
        return call("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Object getBalance(String owner, String coinType) {
        return call("suix_getBalance", List.of(owner, coinType));
    }

    public Object getAllBalances(String owner) {
        return call("suix_getAllBalances", List.of(owner));
    }

    public Object getAllCoins(String owner, String cursor, long limit) {
        return call("suix_getAllCoins", List.of(owner, cursor, limit));
    }

    public Object getCoinMetadata(String coinType) {
        return call("suix_getCoinMetadata", List.of(coinType));
    }

    public Object getCurrentSystemState() {
        return call("suix_getLatestSuiSystemState", List.of());
    }

    public Object getChainIdentifier() {
        return call("sui_getChainIdentifier", List.of());
    }

    public Object getReferenceGasPrice() {
        return call("suix_getReferenceGasPrice", List.of());
    }

    public Object getOwnedObjects(String owner, Map<String, Object> query, Object cursor, long limit) {
        return call("suix_getOwnedObjects", List.of(owner, query, cursor, limit));
    }

    public Object getDynamicFields(String parentId, Object cursor, long limit) {
        return call("suix_getDynamicFields", List.of(parentId, cursor, limit));
    }

    public Object getDynamicFieldObject(String parentId, Map<String, Object> name) {
        return call("suix_getDynamicFieldObject", List.of(parentId, name));
    }

    public Object getTransactionBlock(String digest, Map<String, Object> options) {
        return call("sui_getTransactionBlock", List.of(digest, options));
    }

    public Object multiGetTransactionBlocks(String[] digests, Map<String, Object> options) {
        List<String> digestList = Arrays.stream(digests).filter(Objects::nonNull).collect(Collectors.toList());
        return call("sui_multiGetTransactionBlocks", List.of(digestList, options));
    }

    public Object queryTransactionBlocks(Object query, Object cursor, long limit, boolean descendingOrder) {
        return call("suix_queryTransactionBlocks", List.of(query, cursor, limit, descendingOrder));
    }

    public Object getCheckpoint(String checkpointId) {
        return call("sui_getCheckpoint", List.of(checkpointId));
    }

    public Object getCheckpoints(Object cursor, long limit, boolean descendingOrder) {
        return call("sui_getCheckpoints", List.of(cursor, limit, descendingOrder));
    }

    public Object getCommitteeInfo(String epoch) {
        return call("suix_getCommitteeInfo", List.of(epoch));
    }

    public Object getProtocolConfig(String version) {
        return call("sui_getProtocolConfig", List.of(version));
    }

    public Object resolveNameServiceAddress(String name) {
        return call("suix_resolveNameServiceAddress", List.of(name));
    }

    public Object resolveNameServiceNames(String address, Object cursor, Long limit) {
        return call("suix_resolveNameServiceNames", List.of(address, cursor, limit));
    }

    public Object getEventsByTransaction(String txDigest) {
        return call("sui_getEvents", List.of(txDigest));
    }

    @Override
    public void close() {
        transport.close();
    }
}

File: kotlin-sdks/src/main/kotlin/com/suisdks/sui/grpc/SuiGrpcClient.kt
package com.suisdks.sui.grpc

import java.util.Base64

class SuiGrpcClient(private val transport: GrpcTransport) : AutoCloseable {
    companion object {
        fun fromOfficialGrpc(
            target: String,
            plaintext: Boolean = false,
            security: GrpcSecurityOptions = GrpcSecurityOptions(),
            tls: GrpcTlsOptions = GrpcTlsOptions(),
        ): SuiGrpcClient {
            return SuiGrpcClient(OfficialGrpcTransport(target, plaintext, security, tls))
        }
    }

    fun call(method: String, params: List<Any?> = emptyList(), metadata: Map<String, String> = emptyMap()): Any? {
        val response = transport.unary(GrpcRequest(method = method, params = params, metadata = metadata))
        if (response.hasError()) {
            throw IllegalStateException("gRPC error: ${response.error}")
        }
        return response.result
    }

    fun unary(request: GrpcRequest): GrpcResponse {
        val result = call(request.method, request.params, request.metadata)
        return GrpcResponse(result = result)
    }

    fun batch(requests: List<GrpcRequest>): List<GrpcResponse> = requests.map { request ->
        val result = call(request.method, request.params, request.metadata)
        GrpcResponse(result = result)
    }

    fun discoverRpcApi(): Any? = call("rpc.discover")

    fun getLatestCheckpointSequenceNumber(): Any? = call("sui_getLatestCheckpointSequenceNumber")

    fun getLatestSuiSystemState(): Any? = call("suix_getLatestSuiSystemState")

    fun getCoins(owner: String, coinType: String = "0x2::sui::SUI", cursor: String? = null, limit: Int? = null): Any? =
        call("suix_getCoins", listOf(owner, coinType, cursor, limit))

    fun getAllCoins(owner: String, cursor: String? = null, limit: Int? = null): Any? =
        call("suix_getAllCoins", listOf(owner, cursor, limit))

    fun getBalance(owner: String, coinType: String = "0x2::sui::SUI"): Any? =
        call("suix_getBalance", listOf(owner, coinType))

    fun getAllBalances(owner: String): Any? = call("suix_getAllBalances", listOf(owner))

    fun getCoinMetadata(coinType: String): Any? = call("suix_getCoinMetadata", listOf(coinType))

    fun getTotalSupply(coinType: String): Any? = call("suix_getTotalSupply", listOf(coinType))

    fun getOwnedObjects(owner: String, query: Map<String, Any?> = emptyMap(), cursor: String? = null, limit: Int? = null): Any? =
        call("suix_getOwnedObjects", listOf(owner, query, cursor, limit))

    fun getDynamicFields(parentObjectId: String, cursor: String? = null, limit: Int? = null): Any? =
        call("suix_getDynamicFields", listOf(parentObjectId, cursor, limit))

    fun getDynamicFieldObject(parentObjectId: String, name: Map<String, Any?>): Any? =
        call("suix_getDynamicFieldObject", listOf(parentObjectId, name))

    fun queryEvents(
        query: Map<String, Any?>,
        cursor: String? = null,
        limit: Int? = null,
        descendingOrder: Boolean = false,
    ): Any? = call("suix_queryEvents", listOf(query, cursor, limit, descendingOrder))

    fun queryTransactionBlocks(
        query: Map<String, Any?>,
        cursor: String? = null,
        limit: Int? = null,
        descendingOrder: Boolean = false,
    ): Any? = call("suix_queryTransactionBlocks", listOf(query, cursor, limit, descendingOrder))

    fun getObject(objectId: String, options: Map<String, Any?> = emptyMap()): Any? =
        call("sui_getObject", listOf(objectId, options))

    fun multiGetObjects(objectIds: List<String>, options: Map<String, Any?> = emptyMap()): Any? =
        call("sui_multiGetObjects", listOf(objectIds, options))

    fun getObjects(objectIds: List<String>, options: Map<String, Any?> = emptyMap()): List<GrpcResponse> =
        batch(objectIds.map { GrpcRequest(method = "sui_getObject", params = listOf(it, options)) })

    fun dryRunTransactionBlock(txBytesBase64: String): Any? =
        call("sui_dryRunTransactionBlock", listOf(txBytesBase64))

    fun devInspectTransactionBlock(sender: String, txBytesBase64: String): Any? =
        call("sui_devInspectTransactionBlock", listOf(sender, txBytesBase64))

    fun getEventsByTransaction(transactionDigest: String): Any? =
        call("sui_getEvents", listOf(transactionDigest))

    fun getTransactionBlock(digest: String, options: Map<String, Any?> = emptyMap()): Any? =
        call("sui_getTransactionBlock", listOf(digest, options))

    fun getTotalTransactionBlocks(): Any? = call("sui_getTotalTransactionBlocks")

    fun multiGetTransactionBlocks(digests: List<String>, options: Map<String, Any?> = emptyMap()): Any? =
        call("sui_multiGetTransactionBlocks", listOf(digests, options))

    fun getCheckpoints(cursor: String? = null, limit: Int? = null, descendingOrder: Boolean = false): Any? =
        call("sui_getCheckpoints", listOf(cursor, limit, descendingOrder))

    fun getCheckpoint(checkpointId: String): Any? = call("sui_getCheckpoint", listOf(checkpointId))

    fun getChainIdentifier(): Any? = call("sui_getChainIdentifier")

    fun getCommitteeInfo(epoch: String? = null): Any? = call("suix_getCommitteeInfo", listOf(epoch))

    fun getProtocolConfig(version: String? = null): Any? = call("sui_getProtocolConfig", listOf(version))

    fun getNetworkMetrics(): Any? = call("suix_getNetworkMetrics")

    fun getAddressMetrics(): Any? = call("suix_getLatestAddressMetrics")

    fun getEpochMetrics(cursor: String? = null, limit: Int? = null, descendingOrder: Boolean = false): Any? =
        call("suix_getEpochMetrics", listOf(cursor, limit, descendingOrder))

    fun getAllEpochAddressMetrics(descendingOrder: Boolean = false): Any? =
        call("suix_getAllEpochAddressMetrics", listOf(descendingOrder))

    fun getEpochs(cursor: String? = null, limit: Int? = null, descendingOrder: Boolean = false): Any? =
        call("suix_getEpochs", listOf(cursor, limit, descendingOrder))

    fun getMoveCallMetrics(): Any? = call("suix_getMoveCallMetrics")

    fun getCurrentEpoch(): Any? = call("suix_getCurrentEpoch")

    fun resolveNameServiceAddress(name: String): Any? = call("suix_resolveNameServiceAddress", listOf(name))

    fun resolveNameServiceNames(address: String, cursor: String? = null, limit: Int? = null): Any? =
        call("suix_resolveNameServiceNames", listOf(address, cursor, limit))

    fun getValidatorsApy(): Any? = call("suix_getValidatorsApy")

    fun getStakes(owner: String): Any? = call("suix_getStakes", listOf(owner))

    fun getStakesByIds(stakedSuiIds: List<String>): Any? = call("suix_getStakesByIds", listOf(stakedSuiIds))

    fun tryGetPastObject(objectId: String, version: Int, options: Map<String, Any?> = emptyMap()): Any? =
        call("sui_tryGetPastObject", listOf(objectId, version, options))

    fun tryMultiGetPastObjects(pastObjects: List<Map<String, Any?>>, options: Map<String, Any?> = emptyMap()): Any? =
        call("sui_tryMultiGetPastObjects", listOf(pastObjects, options))

    fun getNormalizedMoveModulesByPackage(packageId: String): Any? =
        call("sui_getNormalizedMoveModulesByPackage", listOf(packageId))

    fun getNormalizedMoveModule(packageId: String, moduleName: String): Any? =
        call("sui_getNormalizedMoveModule", listOf(packageId, moduleName))

    fun getNormalizedMoveFunction(packageId: String, moduleName: String, functionName: String): Any? =
        call("sui_getNormalizedMoveFunction", listOf(packageId, moduleName, functionName))

    fun getMoveFunctionArgTypes(packageId: String, moduleName: String, functionName: String): Any? =
        call("sui_getMoveFunctionArgTypes", listOf(packageId, moduleName, functionName))

    fun getNormalizedMoveStruct(packageId: String, moduleName: String, structName: String): Any? =
        call("sui_getNormalizedMoveStruct", listOf(packageId, moduleName, structName))

    fun verifyZkLoginSignature(bytes: String, signature: String, intentScope: Int, author: String): Any? =
        call("sui_verifyZkLoginSignature", listOf(bytes, signature, intentScope, author))

    fun getReferenceGasPrice(): Any? = call("suix_getReferenceGasPrice")

    fun executeTransactionBlock(
        txBytesBase64: String,
        signatures: List<String> = emptyList(),
        options: Map<String, Any?> = emptyMap(),
    ): Any? = call("sui_executeTransactionBlock", listOf(txBytesBase64, signatures, options))

    fun executeTransactionBlock(
        txBytesBase64: String,
        signature: String,
        options: Map<String, Any?> = emptyMap(),
    ): Any? = executeTransactionBlock(txBytesBase64, listOf(signature), options)

    fun executeTransactionBlock(
        txBytes: ByteArray,
        signatures: List<String> = emptyList(),
        options: Map<String, Any?> = emptyMap(),
    ): Any? = executeTransactionBlock(Base64.getEncoder().encodeToString(txBytes), signatures, options)

    fun executeTransactionBlock(
        txBytes: ByteArray,
        signature: String,
        options: Map<String, Any?> = emptyMap(),
    ): Any? = executeTransactionBlock(txBytes, listOf(signature), options)

    override fun close() {
        transport.close()
    }
}

typealias GrpcCoreClient = SuiGrpcClient
